server/src-lib/Hasura/Server/Migrate.hs

```diff
@@ -64,6 +68,12 @@ instance ToEngineLog MigrationResult Hasura where
             <> latestCatalogVersionString <> "."
     }
 
+-- A migration and (hopefully) also its inverse if we have it.
+data MigrationPair m = MigrationPair
+  { mpMigrate :: m ()
+  , mpDown :: Maybe (m ())
```

Comment:
I don't know why it's polymorphic. It's called by a polymorphic function, with `(MonadIO m, MonadTx m)` as the constraint on `m`, so it needs to be polymorphic here too.

Reply:
I’m not sure I understand what you’re saying—it sounds like you *do* know why it’s polymorphic? But my suggestion was really more about making it clearer why we’re representing a “migration” as an action rather than a piece of data.

Reply:
I see, I thought you were asking why it was using `m ()` rather than some concrete `m`. I only kept it the same as it was before, and I don't know why it used a monadic action before. If we want to change it, I can, but I suspect the only thing to add as a comment is that it was probably an oversight to begin with.

Reply:
I think the answer is that not all of the migrations are actually plain SQL queries: the `from3To4` migration includes some custom Haskell logic. Of course, it seems highly unlikely to me that anyone is actually running a Hasura instance with catalog version *3!* So maybe it makes sense to just drop support for migrating from those catalog versions to simplify the code (it would also eliminate the odd `0.8` special case).

Reply:
I remembered why I ended up writing the recursion. `dropWhile` doesn't quite work here - we need "`dropWhileInclusive`", since we need to include the migration for the version we want itself. Since I'd probably end up writing that via direct recursion, I figured why not just write the full thing directly. Anyway, I've decided to change the approach here to something simpler entirely, and I won't need this function in the new version.